processing/src/main/java/org/apache/druid/segment/nested/NestedFieldLiteralColumnIndexSupplier.java

```diff
@@ -223,16 +214,41 @@ private <T> IntIntPair getLocalRangeFromDictionary(
       }
     }
     globalEndIndex = Math.max(globalStartIndex, globalEndIndex);
-    // end index is not inclusive, so we find the last value in the local dictionary that falls within the range
+
+    // with global dictionary id range settled, now lets map that onto a local dictionary id range
+    int localFound = localDictionary.indexOf(globalStartIndex);
+    if (localFound < 0) {
+      // the first valid global index is not within the local dictionary, so the insertion point is where we begin
+      localStartIndex = -(localFound + 1);
+      // if the computed local start index violates the global range, shift up by 1
+      int actualGlobalStartIndex = localDictionary.get(localStartIndex);
+      if (actualGlobalStartIndex < globalStartIndex) {
```

Comment:
Should this be done in a loop or can we be sure that the global index at the next `localStartIndex` would be valid?

Reply:
ok so i looked a lot closer in debugger and working things out by hand, and this code on the start index is pointless and only can be hit in cases where the range will effectively be empty anyway so can be dropped. it also uncovered a missing bounds check on `FixedIndexed` get method, which I have fixed and added tests for 😅

I've re-arranged some things so the logic is simplified, it was really only the end index that needed adjustment when it was missing from the local dictionary, and that was because I was doing the shifting to account for the end index being exclusive in a funny manner. I think its clearer now. Thanks for asking this question!

Reply:
Thanks! Simpler to follow now.

I also wonder if the last value returned from `FixedIndex.indexOf()` should always be `-(minIndex + 1)`.
My concern is that if we haven't found the value until now, then the insertion point could either be before or after the `currIndex` that we are evaluating. I think we should incorporate that fact into the returned value.

Say at the last step, we found that our `value` is bigger than the `currValue`, so `minIndex` becomes `currIndex + 1`, then we break out of the loop and then we return `-(minIndex + 1) = -(currIndex + 2)`.

Doesn't this mean that we have actually skipped the insertion point? Or maybe I am missing something.
(Although, I guess this should be okay because the `Indexed` interface doesn't make any promises about the returned negative value.)

Reply:
`FixedIndexed.indexOf` is the same-ish implementation as `GenericIndexed.indexOf`, both of which are basically the same as `Arrays.binarySearch`, so I think it is ok. Btw, range finder method is used for both `FixedIndexed` and `GenericIndexed` global dictionaries depending on if it is making `LexicographicalRangeIndex` for nested strings or `NumericRangeIndex` for nested numbers, though the local dictionary is always a `FixedIndexed`.

Or did you mean something about how the range calculation is using `indexOf` to compute the ranges?

btw, I did recently strengthen the contract of `indexOf` to be `(-(insertion point) - 1)` for missing values if `isSorted()` returns true, since these index things basically require the `(-(insertion point) - 1)` behavior to work correctly.

Reply:
Makes sense, thanks for the clarification!